File: clients/Cpp/model/AttackProperties.cpp
#include "AttackProperties.hpp"

AttackProperties::AttackProperties() { }
AttackProperties::AttackProperties(int attackRange, int damage, bool collectResource) : attackRange(attackRange), damage(damage), collectResource(collectResource) { }
AttackProperties AttackProperties::readFrom(InputStream& stream) {
    AttackProperties result;
    result.attackRange = stream.readInt();
    result.damage = stream.readInt();
    result.collectResource = stream.readBool();
    return result;
}
void AttackProperties::writeTo(OutputStream& stream) const {
    stream.write(attackRange);
    stream.write(damage);
    stream.write(collectResource);
}
bool AttackProperties::operator ==(const AttackProperties& other) const {
    return attackRange == other.attackRange && damage == other.damage && collectResource == other.collectResource;
}
size_t std::hash<AttackProperties>::operator ()(const AttackProperties& value) const {
    size_t result = 0;
    result ^= std::hash<int>{}(value.attackRange) + 0x9e3779b9 + (result<<6) + (result>>2);
    result ^= std::hash<int>{}(value.damage) + 0x9e3779b9 + (result<<6) + (result>>2);
    result ^= std::hash<bool>{}(value.collectResource) + 0x9e3779b9 + (result<<6) + (result>>2);
    return result;
}

File: clients/Go/model/attack_properties.go
package model

import "io"
import . "aicup2020/stream"

type AttackProperties struct {
    AttackRange int32
    Damage int32
    CollectResource bool
}
func NewAttackProperties(attackRange int32, damage int32, collectResource bool) AttackProperties {
    return AttackProperties {
        AttackRange: attackRange,
        Damage: damage,
        CollectResource: collectResource,
    }
}
func ReadAttackProperties(reader io.Reader) AttackProperties {
    result := AttackProperties {}
    result.AttackRange = ReadInt32(reader)
    result.Damage = ReadInt32(reader)
    result.CollectResource = ReadBool(reader)
    return result
}
func (value AttackProperties) Write(writer io.Writer) {
    WriteInt32(writer, value.AttackRange)
    WriteInt32(writer, value.Damage)
    WriteBool(writer, value.CollectResource)
}
